File: server/java/render/src/main/java/com/metaobjects/render/recover/XmlForgivingReader.java
package com.metaobjects.render.recover;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Stage-4 tolerant XML reader for the bounded corpus malformation set. Never throws. */
public final class XmlForgivingReader {

    public Map<String, Object> read(String span, boolean caseInsensitive) {
        Map<String, Object> out = new LinkedHashMap<>();
        if (span == null || span.isBlank()) return out;
        int gt = span.indexOf('>');
        if (gt < 0) return out;
        int rootEnd = span.lastIndexOf("</");
        String inner = span.substring(gt + 1, (rootEnd < 0 || rootEnd <= gt) ? span.length() : rootEnd);
        parseChildren(inner, caseInsensitive, out);
        return out;
    }

    private void parseChildren(String inner, boolean ci, Map<String, Object> out) {
        Pattern openTag = Pattern.compile("<([A-Za-z_][A-Za-z0-9_]*)(\\s[^>]*)?>", ci ? Pattern.CASE_INSENSITIVE : 0);
        Matcher m = openTag.matcher(inner);
        int pos = 0;
        while (m.find(pos)) {
            String tag = m.group(1);
            String key = ci ? tag.toLowerCase() : tag;
            int contentStart = m.end();
            String closeRe = "</" + Pattern.quote(tag) + "\\s*>";
            Matcher close = Pattern.compile(closeRe, ci ? Pattern.CASE_INSENSITIVE : 0).matcher(inner);
            int contentEnd, next;
            if (close.find(contentStart)) {
                contentEnd = close.start();
                next = close.end();
            } else {
                // unclosed tag: recover text up to the next sibling open tag
                Matcher sib = openTag.matcher(inner);
                if (sib.find(contentStart)) {
                    contentEnd = sib.start();
                    next = contentEnd;
                } else {
                    contentEnd = inner.length();
                    next = inner.length();
                }
            }
            String content = inner.substring(contentStart, contentEnd);
            Object value = content.contains("<")
                    ? nestedOrText(content, ci)
                    : content.trim();
            accumulate(out, key, value);
            pos = next;
        }
    }

    private Object nestedOrText(String content, boolean ci) {
        Map<String, Object> nested = new LinkedHashMap<>();
        parseChildren(content, ci, nested);
        return nested.isEmpty() ? content.trim() : nested;
    }

    @SuppressWarnings("unchecked")
    private void accumulate(Map<String, Object> out, String key, Object value) {
        if (!out.containsKey(key)) {
            out.put(key, value);
            return;
        }
        Object existing = out.get(key);
        if (existing instanceof List) {
            ((List<Object>) existing).add(value);
        } else {
            List<Object> list = new ArrayList<>();
            list.add(existing);
            list.add(value);
            out.put(key, list);
        }
    }
}

File: server/csharp/MetaObjects.Render/Recover/XmlForgivingReader.cs
using System.Text.RegularExpressions;

namespace MetaObjects.Render.Recover;

/// <summary>
/// Stage-4 tolerant XML reader for the bounded corpus malformation set. Never throws.
/// </summary>
public sealed class XmlForgivingReader
{
    /// <summary>
    /// Parses <paramref name="span"/> as a forgiving XML document, returning
    /// the direct children of the root element as a dictionary keyed by tag name
    /// (lower-cased when <paramref name="caseInsensitive"/> is <c>true</c>).
    /// Values are <c>string</c> for text-only children, a nested
    /// <see cref="Dictionary{String, Object}"/> for element children, or a
    /// <see cref="List{Object}"/> when the same tag name appears more than once.
    /// Returns an empty dictionary for <c>null</c>, blank, or unrecoverable input.
    /// Never throws.
    /// </summary>
    public Dictionary<string, object?> Read(string? span, bool caseInsensitive)
    {
        var out_ = new Dictionary<string, object?>();
        if (string.IsNullOrWhiteSpace(span)) return out_;

        int gt = span.IndexOf('>');
        if (gt < 0) return out_;

        int rootEnd = span.LastIndexOf("</");
        int innerEnd = (rootEnd < 0 || rootEnd <= gt) ? span.Length : rootEnd;
        string inner = span.Substring(gt + 1, innerEnd - (gt + 1));

        ParseChildren(inner, caseInsensitive, out_);
        return out_;
    }

    private static void ParseChildren(string inner, bool ci, Dictionary<string, object?> out_)
    {
        var openTag = new Regex(
            @"<([A-Za-z_][A-Za-z0-9_]*)(\s[^>]*)?>",
            ci ? RegexOptions.IgnoreCase : RegexOptions.None);

        int pos = 0;
        Match m = openTag.Match(inner, pos);
        while (m.Success)
        {
            string tag = m.Groups[1].Value;
            string key = ci ? tag.ToLowerInvariant() : tag;
            int contentStart = m.Index + m.Length;

            string closeRe = @"</" + Regex.Escape(tag) + @"\s*>";
            var closeRegex = new Regex(closeRe, ci ? RegexOptions.IgnoreCase : RegexOptions.None);
            Match close = closeRegex.Match(inner, contentStart);

            int contentEnd, next;
            if (close.Success)
            {
                contentEnd = close.Index;
                next = close.Index + close.Length;
            }
            else
            {
                // unclosed tag: recover text up to the next sibling open tag
                Match sib = openTag.Match(inner, contentStart);
                if (sib.Success)
                {
                    contentEnd = sib.Index;
                    next = contentEnd;
                }
                else
                {
                    contentEnd = inner.Length;
                    next = inner.Length;
                }
            }

            string content = inner.Substring(contentStart, contentEnd - contentStart);
            object? value = content.Contains('<')
                ? NestedOrText(content, ci)
                : content.Trim();

            Accumulate(out_, key, value);
            pos = next;
            if (pos >= inner.Length) break;
            m = openTag.Match(inner, pos);
        }
    }

    private static object? NestedOrText(string content, bool ci)
    {
        var nested = new Dictionary<string, object?>();
        ParseChildren(content, ci, nested);
        return nested.Count == 0 ? content.Trim() : (object?)nested;
    }

    private static void Accumulate(Dictionary<string, object?> out_, string key, object? value)
    {
        if (!out_.ContainsKey(key))
        {
            out_[key] = value;
            return;
        }
        object? existing = out_[key];
        if (existing is List<object?> list)
        {
            list.Add(value);
        }
        else
        {
            var newList = new List<object?> { existing, value };
            out_[key] = newList;
        }
    }
}
